### ISSUE TITLE
Support .tobytes() method on torch.Tensor

### ISSUE THREAD

**Issue by pritamdamania87** [pytorch/pytorch#19143]
## 🚀 Feature
Support .tobytes() method on torch.Tensor

## Motivation

If I need to extract the raw bytes from a Tensor, I need to convert to numpy first and then use `tobytes`. It would be nice to have a tobytes() method for the Tensors themselves. For example, if I have a jpg image stored in a ByteTensor and I would like to open it using PIL, I need to do something like this currently:

```
data = tensor.numpy().tobytes()
img = Image.open(BytesIO(data), mode='r')
```

**Comment by apaszke** [comment]
Note that `tobytes` is very inefficient, as it needs to copy the data of your tensor. A better way would be to implement the buffer interface for tensors, and use `memoryview`.

**Comment by zhangguanheng66** [comment]
Thanks for the support, apaszke. I just edit the title of the issue and label it as a feature request.

**Comment by vadimkantorov** [comment]
Supporting this natively could ease the problem with extremely slow scalar indexing in PyTorch (it also always allocates a temporary Python tensor scalar object before taking item() or float() cast): https://github.com/pytorch/pytorch/issues/29973

**Comment by FirefoxMetzger** [comment]
Anything wrong with `memoryview(tensor.numpy())` or `tensor.numpy().data`? According to [the docs](https://pytorch.org/docs/stable/generated/torch.Tensor.numpy.html) a torch tensor and its numpy array share the data buffer. (I guess exceptions apply if the tensor lives on the GPU?) Since `ndarray.data` or `memoryview(ndarray)` return a `memoryview` into the array's underlying buffer, this will be the exact same buffer. Perhaps I misunderstand the aim here?

On the note of buffers, there is no need to get either a `bytes` or `memoryview` object to construct a PIL image from a tensor. You can simply use `Image.fromarray(tensor.numpy(), mode=...)`. Unless, ofc, you mean by "a jpg image stored in a ByteTensor" that the tensor contains the actual raw/encoded image (interesting use-case ... but why?). In this case, something simple like

```
import imageio as iio
import matplotlib.pyplot as plt

img_decoded = iio.imread(tensor.numpy().data, format="JPEG")
plt.imshow(img_decoded)
```

should do the trick just fine. However, again, I find it hard to come up with a situation where you would actually want to put raw JPEG bytes into a tensor...

**Comment by t-vi** [comment]
Not all cpu tensors can be converted to numpy array because of the dtypes, so there is something wrong with it (but we would need to have a plan for the "format" descriptor for things like bfloat16).

**Comment by FirefoxMetzger** [comment]
@t-vi Fair point 😄 I never thought of bfloat16 outside of the GPU. 

Since there is no native support for bfloat16 in numpy, would it be an idea to use `dtype="V2"` as a fallback type? I don't think this is possible right now, but this way we can expose the buffer as a numpy array similar to supported types. Other applications with bfloat16 support could then take the buffer and cast it back to bfloat16.

It comes with the downside of losing the information that the buffer was bfloat16, so it would become the user's responsibility to track that piece of information manually. It does seem like a natural extension of what's currently possible though, so perhaps it can be a first step?

**Comment by t-vi** [comment]
My take would be to 
- when the user doesn't want the field to be left out nor a byte cast, error,
- either leave out the type field (providing an option to not have it, this is forseen by the buffer protocol) if applications like writing to a file work with that, or cast to a byte stream,
Quantized has a similar problem, in particular with the sub-byte dtypes (quint2x4 and quint4x2).

**Comment by vadimkantorov** [comment]
The other side of the feature request would be to be able to construct a tensor (both no-copy and copy variants may make sense) from a memoryview and from a buffer protocol

**Comment by t-vi** [comment]
Yeah, well, that's relatively easy to achieve using from_blob. I'd do non-copy by default (the memoryview does cater for having a deleter-like mechanism that you would use in that).

I do have an an internal quick and dirty branch of that for my own usecase, I'd have to see if I can open-source it.

**Comment by tillahoffmann** [comment]
Having access to the memoryview interface would also be useful for quickly populating tensors in cython. E.g., consider the following, where `some_complex_iterator` is a placeholder for whatever iteration one might want to do (such as nested loops or conversion from another data structure).

```python
def populate_tensor(tensor: th.Tensor):
    # Python loops are slow.
    for i, x in some_complex_iterator:
        tensor[i] = x
```

Converting the above to cython would allow for quick iteration.

```python
from cython import numeric

def populate_tensor(tensor: th.Tensor):
    cdef numeric[:] buffer = tensor
    # Cython loops are fast.
    for i, x in some_complex_iterator:
        buffer[i] = x
```
